File: backend/src/main/java/com/erp/modules/products/service/ProductService.java
package com.erp.modules.products.service;

import com.erp.modules.products.domain.dto.AddBarcodeRequest;
import com.erp.modules.products.domain.dto.AddComponentRequest;
import com.erp.modules.products.domain.dto.AssignProductBranchRequest;
import com.erp.modules.products.domain.dto.CreateBulkPackRequest;
import com.erp.modules.products.domain.dto.CreateProductRequest;
import com.erp.modules.products.domain.dto.ProductBarcodeDto;
import com.erp.modules.products.domain.dto.ProductBranchDto;
import com.erp.modules.products.domain.dto.ProductBulkPackDto;
import com.erp.modules.products.domain.dto.ProductComponentDto;
import com.erp.modules.products.domain.dto.ProductDto;
import com.erp.modules.products.domain.dto.SetProductPriceRequest;
import com.erp.modules.products.domain.dto.SetProductWeighingRequest;
import com.erp.modules.products.domain.dto.ProductPriceDto;
import com.erp.modules.products.domain.dto.UnitOfMeasureDto;
import com.erp.modules.products.domain.dto.UpdateProductRequest;
import java.util.List;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

public interface ProductService {

    ProductDto create(CreateProductRequest req);

    ProductDto getByUid(String uid);

    /**
     * Internal lookup by numeric id — for cross-module callers (e.g. stock-count snapshot)
     * that hold a productId and need code/name for denormalisation.
     * No scope check; caller is responsible for asserting company scope.
     */
    ProductDto getById(Long id);

    Page<ProductDto> list(Long companyId, String q, Pageable pageable);

    ProductDto updateByUid(String uid, UpdateProductRequest req);

    void archiveByUid(String uid);

    void restoreByUid(String uid);

    // --- Branches ---
    ProductBranchDto assignBranch(String uid, AssignProductBranchRequest req);

    void removeBranch(String uid, String branchUid);

    List<ProductBranchDto> listBranches(String uid);

    // --- Bulk packs ---
    ProductBulkPackDto addBulkPack(String uid, CreateBulkPackRequest req);

    void removeBulkPack(String productUid, String bulkPackUid);

    List<ProductBulkPackDto> listBulkPacks(String uid);

    // --- Barcodes ---
    ProductBarcodeDto addBarcode(String uid, AddBarcodeRequest req);

    void removeBarcode(String productUid, String barcodeUid);

    List<ProductBarcodeDto> listBarcodes(String uid);

    ProductBarcodeDto lookupBarcode(Long companyId, String barcode);

    // --- Prices ---
    ProductPriceDto setPrice(String uid, SetProductPriceRequest req);

    /**
     * Removes a price row. {@code unitUid} null/blank targets the base-unit row (unit_id IS NULL,
     * back-compatible with pre-ADR-0048 callers); a supplied uid targets that unit's per-unit row
     * (coerced to the base row when the uid resolves to the product's own base unit).
     */
    void removePrice(String productUid, String priceListUid, String unitUid);

    List<ProductPriceDto> listPrices(String uid);

    /**
     * Configures the product's weighed-goods (sell-by-weight) profile (ADR-0044 D-1b). When
     * {@code weighed=true} the product's base unit must be a WEIGHT unit, else the call is rejected.
     */
    ProductDto setWeighing(String uid, SetProductWeighingRequest req);

    // --- Valid transaction units ---

    /**
     * Returns the ordered list of units valid for transaction lines against this product:
     * the product's base unit first, then each configured ProductBulkPack unit (de-duplicated).
     * Used by the /uid/{uid}/units endpoint to drive unit-of-measure pickers in line editors.
     */
    List<UnitOfMeasureDto> listProductUnits(String productUid);

    // --- Components ---
    ProductComponentDto addComponent(String uid, AddComponentRequest req);

    void removeComponent(String composedUid, String componentUid);

    List<ProductComponentDto> listComponents(String uid);
}

File: web/src/app/features/admin/products/product.service.ts
import { HttpClient, HttpContext, HttpParams } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable, map } from 'rxjs';
import { ApiResponse, PageMeta } from '../../../core/api/api-response.model';
import { SKIP_UNWRAP } from '../../../core/api/http-context.tokens';
import { environment } from '../../../../environments/environment';
import {
  AddBarcodeRequest,
  AddComponentRequest,
  AssignProductBranchRequest,
  CreateBulkPackRequest,
  CreatePriceListRequest,
  CreateProductRequest,
  CreateUnitOfMeasureRequest,
  PriceListDto,
  ProductBarcodeDto,
  ProductBranchDto,
  ProductBulkPackDto,
  ProductComponentDto,
  ProductModel,
  ProductPriceDto,
  SetProductPriceRequest,
  UnitOfMeasureDto,
  UpdatePriceListRequest,
  UpdateProductRequest,
  UpdateUnitOfMeasureRequest,
} from '../models/product.model';

export interface ProductPage {
  rows: ProductModel[];
  meta: PageMeta;
}

/**
 * Product API client.
 * list() uses SKIP_UNWRAP to read both data and PageMeta.
 * All other methods use the auto-unwrap path (interceptor strips the envelope).
 * Base: /api/v1/products; price-lists: /api/v1/price-lists.
 */
@Injectable({ providedIn: 'root' })
export class ProductService {
  private readonly http = inject(HttpClient);
  private readonly base = `${environment.apiBaseUrl}/products`;
  private readonly priceListBase = `${environment.apiBaseUrl}/price-lists`;

  // ── Product CRUD ──────────────────────────────────────────────────────────

  list(companyId: string, q?: string, page = 0, size = 20): Observable<ProductPage> {
    let params = new HttpParams()
      .set('companyId', companyId)
      .set('page', String(page))
      .set('size', String(size));
    if (q?.trim()) params = params.set('q', q.trim());

    const context = new HttpContext().set(SKIP_UNWRAP, true);
    return this.http
      .get<ApiResponse<ProductModel[]>>(this.base, { params, context })
      .pipe(
        map((env) => ({
          rows: env.data ?? [],
          meta: env.meta ?? { page, size, totalElements: env.data?.length ?? 0, totalPages: 1, hasNext: false },
        })),
      );
  }

  getByUid(uid: string): Observable<ProductModel> {
    return this.http.get<ProductModel>(`${this.base}/uid/${uid}`);
  }

  create(request: CreateProductRequest): Observable<ProductModel> {
    return this.http.post<ProductModel>(this.base, request);
  }

  update(uid: string, request: UpdateProductRequest): Observable<ProductModel> {
    return this.http.put<ProductModel>(`${this.base}/uid/${uid}`, request);
  }

  archive(uid: string): Observable<ProductModel> {
    return this.http.put<ProductModel>(`${this.base}/uid/${uid}/archive`, {});
  }

  restore(uid: string): Observable<ProductModel> {
    return this.http.put<ProductModel>(`${this.base}/uid/${uid}/restore`, {});
  }

  barcodeLookup(companyId: string, barcode: string): Observable<ProductModel> {
    const params = new HttpParams().set('companyId', companyId).set('barcode', barcode);
    return this.http.get<ProductModel>(`${this.base}/barcode-lookup`, { params });
  }

  // ── Barcodes ──────────────────────────────────────────────────────────────

  listBarcodes(uid: string): Observable<ProductBarcodeDto[]> {
    return this.http.get<ProductBarcodeDto[]>(`${this.base}/uid/${uid}/barcodes`);
  }

  addBarcode(uid: string, request: AddBarcodeRequest): Observable<ProductBarcodeDto> {
    return this.http.post<ProductBarcodeDto>(`${this.base}/uid/${uid}/barcodes`, request);
  }

  removeBarcode(uid: string, barcodeUid: string): Observable<void> {
    return this.http.delete<void>(`${this.base}/uid/${uid}/barcodes/${barcodeUid}`);
  }

  // ── Bulk Packs ────────────────────────────────────────────────────────────

  listBulkPacks(uid: string): Observable<ProductBulkPackDto[]> {
    return this.http.get<ProductBulkPackDto[]>(`${this.base}/uid/${uid}/bulk-packs`);
  }

  addBulkPack(uid: string, request: CreateBulkPackRequest): Observable<ProductBulkPackDto> {
    return this.http.post<ProductBulkPackDto>(`${this.base}/uid/${uid}/bulk-packs`, request);
  }

  removeBulkPack(uid: string, bulkPackUid: string): Observable<void> {
    return this.http.delete<void>(`${this.base}/uid/${uid}/bulk-packs/${bulkPackUid}`);
  }

  // ── Prices ────────────────────────────────────────────────────────────────

  listPrices(uid: string): Observable<ProductPriceDto[]> {
    return this.http.get<ProductPriceDto[]>(`${this.base}/uid/${uid}/prices`);
  }

  /** unitUid absent/undefined ⇒ base-unit price row; set ⇒ per-unit (pack) price row (ADR-0048). */
  setPrice(uid: string, request: SetProductPriceRequest): Observable<ProductPriceDto> {
    return this.http.post<ProductPriceDto>(`${this.base}/uid/${uid}/prices`, request);
  }

  /** unitUid absent ⇒ base row (back-compatible); set ⇒ the per-unit (pack) price row (ADR-0048). */
  removePrice(uid: string, priceListUid: string, unitUid?: string): Observable<void> {
    const params = unitUid ? new HttpParams().set('unitUid', unitUid) : undefined;
    return this.http.delete<void>(`${this.base}/uid/${uid}/prices/${priceListUid}`, { params });
  }

  // ── Components / Recipe ───────────────────────────────────────────────────

  listComponents(uid: string): Observable<ProductComponentDto[]> {
    return this.http.get<ProductComponentDto[]>(`${this.base}/uid/${uid}/components`);
  }

  addComponent(uid: string, request: AddComponentRequest): Observable<ProductComponentDto> {
    return this.http.post<ProductComponentDto>(`${this.base}/uid/${uid}/components`, request);
  }

  removeComponent(uid: string, componentUid: string): Observable<void> {
    return this.http.delete<void>(`${this.base}/uid/${uid}/components/${componentUid}`);
  }

  // ── Branch Associations ───────────────────────────────────────────────────

  listBranches(uid: string): Observable<ProductBranchDto[]> {
    return this.http.get<ProductBranchDto[]>(`${this.base}/uid/${uid}/branches`);
  }

  assignBranch(uid: string, request: AssignProductBranchRequest): Observable<ProductBranchDto> {
    return this.http.post<ProductBranchDto>(`${this.base}/uid/${uid}/branches`, request);
  }

  removeBranch(uid: string, branchUid: string): Observable<void> {
    return this.http.delete<void>(`${this.base}/uid/${uid}/branches/${branchUid}`);
  }

  // ── Product-scoped units ──────────────────────────────────────────────────

  /**
   * Returns the valid transaction units for a specific product:
   * base unit first, then each active bulk-pack unit (de-duplicated).
   * Calls GET /api/v1/products/uid/{uid}/units — gated by PRODUCT.VIEW.
   */
  listProductUnits(uid: string): Observable<UnitOfMeasureDto[]> {
    return this.http.get<UnitOfMeasureDto[]>(`${this.base}/uid/${uid}/units`);
  }

  // ── Units of Measure ─────────────────────────────────────────────────────

  private readonly unitBase = `${environment.apiBaseUrl}/units`;

  listUnits(companyId: string, q?: string, page = 0, size = 200): Observable<{ rows: UnitOfMeasureDto[]; meta: PageMeta }> {
    let params = new HttpParams()
      .set('companyId', companyId)
      .set('page', String(page))
      .set('size', String(size));
    if (q?.trim()) params = params.set('q', q.trim());
    const context = new HttpContext().set(SKIP_UNWRAP, true);
    return this.http
      .get<ApiResponse<UnitOfMeasureDto[]>>(this.unitBase, { params, context })
      .pipe(
        map((env) => ({
          rows: env.data ?? [],
          meta: env.meta ?? { page, size, totalElements: env.data?.length ?? 0, totalPages: 1, hasNext: false },
        })),
      );
  }

  getUnitByUid(uid: string): Observable<UnitOfMeasureDto> {
    return this.http.get<UnitOfMeasureDto>(`${this.unitBase}/uid/${uid}`);
  }

  createUnit(request: CreateUnitOfMeasureRequest): Observable<UnitOfMeasureDto> {
    return this.http.post<UnitOfMeasureDto>(this.unitBase, request);
  }

  updateUnit(uid: string, request: UpdateUnitOfMeasureRequest): Observable<UnitOfMeasureDto> {
    return this.http.put<UnitOfMeasureDto>(`${this.unitBase}/uid/${uid}`, request);
  }

  archiveUnit(uid: string): Observable<UnitOfMeasureDto> {
    return this.http.put<UnitOfMeasureDto>(`${this.unitBase}/uid/${uid}/archive`, {});
  }

  restoreUnit(uid: string): Observable<UnitOfMeasureDto> {
    return this.http.put<UnitOfMeasureDto>(`${this.unitBase}/uid/${uid}/restore`, {});
  }

  // ── Price Lists ───────────────────────────────────────────────────────────

  listPriceLists(companyId: string): Observable<PriceListDto[]> {
    return this.http.get<PriceListDto[]>(this.priceListBase, { params: { companyId } });
  }

  getPriceListByUid(uid: string): Observable<PriceListDto> {
    return this.http.get<PriceListDto>(`${this.priceListBase}/uid/${uid}`);
  }

  createPriceList(request: CreatePriceListRequest): Observable<PriceListDto> {
    return this.http.post<PriceListDto>(this.priceListBase, request);
  }

  updatePriceList(uid: string, request: UpdatePriceListRequest): Observable<PriceListDto> {
    return this.http.put<PriceListDto>(`${this.priceListBase}/uid/${uid}`, request);
  }

  archivePriceList(uid: string): Observable<PriceListDto> {
    return this.http.put<PriceListDto>(`${this.priceListBase}/uid/${uid}/archive`, {});
  }

  restorePriceList(uid: string): Observable<PriceListDto> {
    return this.http.put<PriceListDto>(`${this.priceListBase}/uid/${uid}/restore`, {});
  }
}
